### PR TITLE
Unique lanelets in Ego route fix

### PR DESCRIPTION
**Pull request by piotr-zyskowski-rai**
## Types of PR

- [ ] New Features
- [ ] Upgrade of existing features
- [x] Bugfix

## Link to the issue
https://tier4.atlassian.net/browse/AJD-600

### CODE REVIEW

**Comment by HansRobo** [comment]
@piotr-zyskowski-rai Thank you for your contribution!
But, first of all, could you make all ci ckecks green?

**Comment by HansRobo** [comment]
After fixing ci errors, please mark this pull request as "ready for review".
Then the review will strat.

File: simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp [hunk 1]
```diff
@@ -111,7 +111,7 @@ class EgoEntity : public VehicleEntity
 
   auto getObstacle() -> boost::optional<traffic_simulator_msgs::msg::Obstacle> override;
 
-  auto getRouteLanelets() const -> std::vector<std::int64_t>;
+  auto getRouteUniqueLanelets() const -> std::vector<std::int64_t>;
```
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp:114]: It is not a good idea to filter at this point, as I believe there can be cases where the loop route is correct.
Also, the order of RouteLanelets is very important, so std::unorderd_set, which automatically rearranges them, is not a good idea.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

**Comment by kielczykowski-rai** [comment]
@hakuturu583 
I took the task over from @piotr-zyskowski-rai, and I implemented the changes you requested. Could you take a look if the implementation meets your expectations? I still work on the proper lifting of the file so the pushed commits are only related to the linter matter.

PS. sorry for the force-pushes - I will avoid them in the future :bowing_man:

File: simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp [hunk 2]
```diff
@@ -74,7 +75,7 @@ class EgoEntity : public VehicleEntity
     -> const std::shared_ptr<SimModelInterface>;
 
 public:
-  explicit EgoEntity() = delete;
```
Comment (hakuturu583) [inline simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp:77]: Why do you remove explicit?

File: simulation/traffic_simulator/include/traffic_simulator/helper/helper.hpp [hunk 3]
```diff
@@ -89,6 +92,26 @@ geometry_msgs::msg::Vector3 constructRPYfromQuaternion(geometry_msgs::msg::Quate
...
   double x, double y, double z, double roll, double pitch, double yaw);
 
+/**
+ * @brief helper function for creating vector without duplicates, with preserved order
+ *
+ * @param vector input std::vector
+ * @return new std::vector without duplicates and with relative order preserved
+ */
+template <typename T>
+std::vector<T> getUniqueValues(std::vector<T> input_vector)
```
Comment (hakuturu583) [inline simulation/traffic_simulator/include/traffic_simulator/helper/helper.hpp:102]: How about using const reference.

```suggestion
std::vector<T> getUniqueValues(const std::vector<T> & input_vector)
```

File: simulation/traffic_simulator/src/entity/ego_entity.cpp [hunk 4]
```diff
@@ -82,17 +82,23 @@ auto EgoEntity::makeSimulationModel(
...
+  const auto vel_lim = getParameter<double>("vel_lim", parameters.performance.max_speed);   // 50.0
+  const auto vel_rate_lim = getParameter<double>(
+    "vel_rate_lim",
+    parameters.performance.max_acceleration);                                               // 7.0
+  const auto vel_time_constant = getParameter<double>("vel_time_constant", 0.1);
+  const auto vel_time_delay = getParameter<double>("vel_time_delay", 0.1);
+  const auto wheel_base = getParameter<double>(
+    "wheel_base",
+    parameters.axles.front_axle.position_x -
+    parameters.axles.rear_axle.position_x);
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/entity/ego_entity.cpp:101]: These codes were intentionally aligned.
Please correct them.

File: simulation/traffic_simulator/src/entity/ego_entity.cpp [hunk 5]
```diff
@@ -207,15 +213,16 @@ auto EgoEntity::getEntityStatus(const double time, const double step_time) const
       return accel;
     }();
 
-    const auto route_lanelets = getRouteLanelets();
+    auto route_lanelets = getRouteLanelets();
+    const auto unique_route_lanelets = traffic_simulator::helper::getUniqueValues(route_lanelets);
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/entity/ego_entity.cpp:217]: route lanelets is passed to getUniqueValues function once.
So, I suggest this  change.

```suggestion
    const auto unique_route_lanelets =
      traffic_simulator::helper::getUniqueValues(getRouteLanelets());
```

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp (hunk 2 above)
Comment (kielczykowski-rai) [reply simulation/traffic_simulator/include/traffic_simulator/entity/ego_entity.hpp]: I changed the line due to some build error, which occurred to me once building the SS2 with Autoware. As I checked now, it builds on humble correctly, so i revoked the change. Thank you

**Comment by kielczykowski-rai** [comment]
@hakuturu583 
Thank you for the review. I have applied the suggestions and built and tested the solution once again.
I will check if CI checks are successful, and then can I test the branch with the evaluator?

**Comment by hakuturu583** [comment]
> I will check if CI jobs are successful, and then can I test the branch with the evaluator?
OK I will review your change soon, please wait a bit.

**Comment by hakuturu583** [comment]
Looks good to me, please try this branch on evaluator.

**Comment by hakuturu583** [comment]
@kielczykowski-rai friendly ping

**Comment by kielczykowski-rai** [comment]
@hakuturu583 
I have run the change in the evaluator. For now, I have discovered other errors related to CI performance. I will leave an update when the regression tests are done.

**Comment by HansRobo** [comment]
@hakuturu583 Michał-san finished regression test and it works fine!
Could you do final-review & merge this pull request?

c.f. https://tier4.atlassian.net/browse/AJD-600

**Review by hakuturu583 — APPROVED** [review]

**Pull request merged**